File: Layout.java
// DynamicLayout Java API — generate layout JSON from pure Java code.
//
// Single-file. Zero dependencies (org.json is excluded — manual JSON building).
// Works on Java 11+.
//
// Example:
//   Layout page = Layout.begin("My Page");
//   page.fieldset("Info").label("Hello from Java!").end();
//   page.button("ok", "OK", "primary", true);
//   System.out.println(page.end());

package dynamiclayout;

import java.util.*;

public class Layout {
    private final String title, uid;
    private final List<Map<String, Object>> layout = new ArrayList<>();
    private final List<Map<String, Object>> actions = new ArrayList<>();
    private final Deque<List<Map<String, Object>>> stack = new ArrayDeque<>();
    private int counter;

    private Layout(String title) {
        this.title = title;
        this.uid = "layout_" + System.currentTimeMillis();
        stack.push(layout);
    }

    public static Layout begin(String title) { return new Layout(title); }
    public Layout uid(String uid) { return this; } // uid set by constructor

    private String k(String p) { return p + "_" + (++counter); }
    private List<Map<String, Object>> top() { return stack.peek(); }
    private Layout add(Map<String, Object> el) { top().add(el); return this; }
    private Layout push(Map<String, Object> el) {
        add(el); List<Map<String, Object>> c = new ArrayList<>();
        el.put("content", c); stack.push(c); return this;
    }
    private Layout pop() { stack.pop(); return this; }

    private static Map<String, Object> el(String type, String key, Object... kv) {
        Map<String, Object> m = new LinkedHashMap<>(); m.put("type", type); m.put("key", key);
        for (int i = 0; i + 1 < kv.length; i += 2) m.put((String) kv[i], kv[i + 1]);
        return m;
    }

    // Containers
    public Layout fieldset(String title) { return push(el("FIELDSET", k("fs"), "title", title)); }
    public Layout fieldsetCollapsed(String title) { return push(el("FIELDSET", k("fs"), "title", title, "collapsed", true)); }
    public Layout end() { return pop(); }
    public Layout row() { return push(el("ROW", k("r"))); }
    public Layout endRow() { return pop(); }
    public Layout col(int... sizes) {
        Map<String, Object> e = el("COL", k("c"));
        if (sizes.length > 0) { Map<String, Object> len = new LinkedHashMap<>();
            String[] keys = {"xs", "sm", "md", "lg", "xl"};
            for (int i = 0; i < Math.min(sizes.length, 5); i++) if (sizes[i] > 0) len.put(keys[i], sizes[i]);
            if (!len.isEmpty()) e.put("length", len);
        }
        return push(e);
    }
    public Layout endCol() { return pop(); }

    // Display
    public Layout label(String text) { return add(el("LABEL", k("l"), "label", text)); }
    public Layout alert(String message, String color) { return add(el("ALERT", k("a"), "message", message, "color", color)); }
    public Layout badge(String title, String color) { return add(el("BADGE", k("bd"), "title", title, "color", color)); }
    public Layout spacer(int px) { return add(el("SPACER", k("sp"), "width", px)); }
    public Layout progress(String label, int pct, String color) {
        Map<String, Object> e = el("PROGRESS", k("pr"), "progress", pct);
        if (label != null && !label.isEmpty()) e.put("label", label);
        if (color != null) e.put("color", color);
        return add(e);
    }

    // Inputs
    public Layout input(String id, String label) { return add(el("INPUT", k("i"), "id", id, "label", label)); }
    public Layout inputRequired(String id, String label) { return add(el("INPUT", k("i"), "id", id, "label", label, "required", true)); }
    public Layout textarea(String id, String label, int rows) { return add(el("TEXTAREA", k("ta"), "id", id, "label", label, "rows", rows)); }
    public Layout checkbox(String id, String label) { return add(el("CHECKBOX", k("cb"), "id", id, "label", label)); }
    public Layout select(String id, String label, String[]... pairs) {
        List<Map<String, Object>> vals = new ArrayList<>();
        for (String[] p : pairs) vals.add(el("", "", "id", p[0], "displayName", p[1]));
        return add(el("SELECT", k("s"), "id", id, "label", label, "values", vals));
    }
    public Layout rating(String id, String label) { return add(el("RATING", k("rt"), "id", id, "label", label)); }
    public Layout readonly(String id, String label) { return add(el("READONLY_FIELD", k("ro"), "id", id, "label", label)); }

    // Actions
    public Layout button(String id, String title, String color, boolean isDefault) {
        Map<String, Object> e = el("BUTTON", k("btn"), "id", id, "title", title, "color", color);
        if (isDefault) e.put("default", true);
        actions.add(e); return this;
    }

    // Shortcuts
    public Layout section(String title, String text) { return fieldset(title).label(text).end(); }
    public Layout feedbackForm() {
        return fieldset("Your Details")
            .inputRequired("name", "Name")
            .inputRequired("email", "Email")
            .textarea("message", "Message", 5)
            .end()
            .button("send", "Send", "primary", true)
            .button("cancel", "Cancel", "secondary", false);
    }

    // Output — manual JSON building
    public String endLayout() {
        while (stack.size() > 1) pop();
        StringBuilder sb = new StringBuilder();
        sb.append("{\n  \"ui\": {\n");
        appendStr(sb, 2, "title", title);
        appendStr(sb, 2, "uid", uid);
        appendArr(sb, 2, "layout", layout);
        if (!actions.isEmpty()) appendArr(sb, 2, "actions", actions);
        sb.append("    \"translations\": {},\n");
        sb.append("    \"userAccess\": { \"cancel\": true }\n");
        sb.append("  }\n}");
        return sb.toString();
    }

    private void appendStr(StringBuilder sb, int indent, String key, String val) {
        sb.append("  ".repeat(indent)).append('"').append(key).append("\": \"")
          .append(val.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n")).append("\",\n");
    }

    private void appendArr(StringBuilder sb, int indent, String key, List<Map<String, Object>> arr) {
        sb.append("  ".repeat(indent)).append('"').append(key).append("\": [\n");
        for (int i = 0; i < arr.size(); i++) {
            appendObj(sb, indent + 2, arr.get(i));
            if (i < arr.size() - 1) sb.append(',');
            sb.append('\n');
        }
        sb.append("  ".repeat(indent)).append("],\n");
    }

    private void appendObj(StringBuilder sb, int indent, Map<String, Object> obj) {
        sb.append("  ".repeat(indent)).append("{\n");
        List<String> keys = new ArrayList<>(obj.keySet());
        for (int i = 0; i < keys.size(); i++) {
            String k = keys.get(i);
            Object v = obj.get(k);
            sb.append("  ".repeat(indent + 1)).append('"').append(k).append("\": ");
            appendVal(sb, indent + 1, v);
            if (i < keys.size() - 1) sb.append(',');
            sb.append('\n');
        }
        sb.append("  ".repeat(indent)).append('}');
    }

    private void appendVal(StringBuilder sb, int indent, Object v) {
        if (v instanceof String) sb.append('"').append(((String) v).replace("\\", "\\\\").replace("\"", "\\\"")).append('"');
        else if (v instanceof Integer) sb.append(v);
        else if (v instanceof Boolean) sb.append(v);
        else if (v instanceof List) {
            sb.append('[');
            List<?> lst = (List<?>) v;
            for (int i = 0; i < lst.size(); i++) {
                if (i > 0) sb.append(", ");
                if (lst.get(i) instanceof Map) { appendObj(sb, 0, (Map<String, Object>) lst.get(i)); }
                else appendVal(sb, indent, lst.get(i));
            }
            sb.append(']');
        }
    }
}

File: Layout.kt
// DynamicLayout Kotlin standalone API — generate layout JSON from Kotlin code.
//
// Single-file. Zero dependencies outside Kotlin stdlib.
// Works on Kotlin 1.8+ (JVM, JS, Native via kotlinx.serialization optional).
//
// Example:
//   val page = Layout("My Page")
//   page.fieldset("Info") { label("Hello from Kotlin!") }
//   page.button("ok", "OK", "primary", isDefault = true)
//   println(page.json())

package dynamiclayout

import kotlin.collections.*

// Types
typealias El = MutableMap<String, Any?>

class Layout(private val title: String, private val uid: String = "layout_${System.currentTimeMillis()}") {
    private val layout = mutableListOf<El>()
    private val actions = mutableListOf<El>()
    private val stack = mutableListOf(layout)
    private var counter = 0

    private fun k(p: String) = "${p}_${++counter}"
    private fun top() = stack.last()
    private fun add(el: El): Layout { top().add(el); return this }

    private fun push(el: El): Layout {
        val content = mutableListOf<El>()
        el["content"] = content; add(el); stack.add(content); return this
    }

    private fun pop(): Layout { stack.removeLast(); return this }

    // Containers
    fun fieldset(title: String, collapsed: Boolean? = null, block: Layout.() -> Unit = {}): Layout {
        val el = mutableMapOf<String, Any?>("type" to "FIELDSET", "key" to k("fs"), "title" to title)
        if (collapsed != null) el["collapsed"] = collapsed
        push(el); block(); pop(); return this
    }

    fun row(block: Layout.() -> Unit): Layout {
        push(mutableMapOf("type" to "ROW", "key" to k("r"))); block(); pop(); return this
    }

    fun col(xs: Int = 0, sm: Int = 0, md: Int = 0, lg: Int = 0, xl: Int = 0, block: Layout.() -> Unit = {}): Layout {
        val el = mutableMapOf<String, Any?>("type" to "COL", "key" to k("c"))
        val len = mutableMapOf<String, Int>()
        if (xs > 0) len["xs"] = xs; if (sm > 0) len["sm"] = sm; if (md > 0) len["md"] = md
        if (lg > 0) len["lg"] = lg; if (xl > 0) len["xl"] = xl
        if (len.isNotEmpty()) el["length"] = len
        push(el); block(); pop(); return this
    }

    fun inline(block: Layout.() -> Unit): Layout {
        push(mutableMapOf("type" to "INLINE_GROUP", "key" to k("ig"))); block(); pop(); return this
    }

    // Display
    fun label(text: String): Layout = add(mutableMapOf("type" to "LABEL", "key" to k("l"), "label" to text))
    fun alert(message: String, color: String = "info"): Layout = add(mutableMapOf("type" to "ALERT", "key" to k("a"), "message" to message, "color" to color))
    fun badge(title: String, color: String = "primary"): Layout = add(mutableMapOf("type" to "BADGE", "key" to k("bd"), "title" to title, "color" to color))
    fun spacer(px: Int = 20): Layout = add(mutableMapOf("type" to "SPACER", "key" to k("sp"), "width" to px))
    fun progress(pct: Int, label: String? = null, color: String? = null): Layout {
        val el = mutableMapOf<String, Any?>("type" to "PROGRESS", "key" to k("pr"), "progress" to pct)
        label?.let { el["label"] = it }; color?.let { el["color"] = it }
        return add(el)
    }

    // Inputs
    fun input(id: String, label: String = "", required: Boolean = false, dataType: String = "STRING", maxLength: Int = 0): Layout {
        val el = mutableMapOf<String, Any?>("type" to "INPUT", "key" to k("i"), "id" to id, "label" to label)
        if (required) el["required"] = true
        if (dataType != "STRING") el["dataType"] = dataType
        if (maxLength > 0) el["maxLength"] = maxLength
        return add(el)
    }
    fun textarea(id: String, label: String = "", rows: Int = 3): Layout = add(mutableMapOf("type" to "TEXTAREA", "key" to k("ta"), "id" to id, "label" to label, "rows" to rows))
    fun checkbox(id: String, label: String = ""): Layout = add(mutableMapOf("type" to "CHECKBOX", "key" to k("cb"), "id" to id, "label" to label))
    fun select(id: String, label: String, vararg pairs: Pair<String, String>): Layout {
        val vals = pairs.map { mapOf("id" to it.first, "displayName" to it.second) }
        return add(mutableMapOf("type" to "SELECT", "key" to k("s"), "id" to id, "label" to label, "values" to vals))
    }
    fun radio(id: String, label: String, vararg pairs: Pair<String, String>): Layout {
        val vals = pairs.map { mapOf("id" to it.first, "displayName" to it.second) }
        return add(mutableMapOf("type" to "RADIOBUTTON", "key" to k("rb"), "id" to id, "label" to label, "values" to vals))
    }
    fun rating(id: String, label: String = ""): Layout = add(mutableMapOf("type" to "RATING", "key" to k("rt"), "id" to id, "label" to label))
    fun readonly(id: String, label: String = ""): Layout = add(mutableMapOf("type" to "READONLY_FIELD", "key" to k("ro"), "id" to id, "label" to label))

    // Actions
    fun button(id: String, title: String, color: String = "primary", isDefault: Boolean = false): Layout {
        val el = mutableMapOf<String, Any?>("type" to "BUTTON", "key" to k("btn"), "id" to id, "title" to title, "color" to color)
        if (isDefault) el["default"] = true
        actions.add(el); return this
    }

    // Shortcuts
    fun section(title: String, text: String): Layout = fieldset(title) { label(text) }
    fun feedbackForm(): Layout {
        fieldset("Your Details") {
            input("name", "Name", required = true)
            input("email", "Email", required = true)
            textarea("message", "Message", 5)
        }
        button("send", "Send", "primary", isDefault = true)
        button("cancel", "Cancel", "secondary")
        return this
    }

    // Output
    fun toMap(): Map<String, Any?> {
        while (stack.size > 1) pop()
        val ui = mutableMapOf<String, Any?>(
            "title" to title, "uid" to uid, "layout" to layout,
            "translations" to emptyMap<String, String>(), "userAccess" to mapOf("cancel" to true)
        )
        if (actions.isNotEmpty()) ui["actions"] = actions
        return mapOf("ui" to ui)
    }

    fun json(): String {
        val sb = StringBuilder()
        appendMap(sb, 0, toMap())
        return sb.toString()
    }

    private fun appendMap(sb: StringBuilder, indent: Int, map: Map<String, Any?>) {
        sb.append("{\n")
        val entries = map.entries.toList()
        entries.forEachIndexed { i, (k, v) ->
            sb.append("  ".repeat(indent + 1))
            sb.append("\"$k\": ")
            appendVal(sb, indent + 1, v)
            if (i < entries.size - 1) sb.append(',')
            sb.append('\n')
        }
        sb.append("  ".repeat(indent)); sb.append('}')
    }

    private fun appendVal(sb: StringBuilder, indent: Int, v: Any?) {
        when (v) {
            null -> sb.append("null")
            is String -> {
                sb.append('"')
                sb.append(v.replace("\\", "\\\\").replace("\"", "\\\"").replace("\n", "\\n"))
                sb.append('"')
            }
            is Number -> sb.append(v)
            is Boolean -> sb.append(v)
            is Map<*, *> -> @Suppress("UNCHECKED_CAST") appendMap(sb, indent, v as Map<String, Any?>)
            is List<*> -> {
                sb.append('[')
                v.forEachIndexed { i, item ->
                    if (i > 0) sb.append(", ")
                    appendVal(sb, indent, item)
                }
                sb.append(']')
            }
            else -> sb.append('"').append(v.toString()).append('"')
        }
    }
}
